### PR TITLE
Sharathkrml/rarible

### PR DESCRIPTION
**Pull request by sharathkrml**

### CODE REVIEW

File: rarible/ExchangeV1.sol [hunk 1]
```diff
@@ -0,0 +1,1494 @@
+/**
+ *Submitted for verification at Etherscan.io on 2020-11-17
+ */
+
+pragma solidity ^0.5.0;
```
Comment (0xsarvesh) [inline rarible/ExchangeV1.sol:5]: Please remove solidity files

File: rarible/data.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
+transaction value = 0.10234597876321875 Ether
+Etherscan detected transaction value = 0.00399398941515 Eth
+buyValue =  0.5990984122725
+amount = 5
```
Comment (0xsarvesh) [inline rarible/data.txt:4]: What is this file? 
is it just for debugging ? lets delete this.

File: rarible/networks.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "mainnet": {
+    "ExchangeV1": {
+      "address": "0xcd4ec7b66fbc029c116ba9ffb3e59351c20b5b06"
```
Comment (0xsarvesh) [inline rarible/networks.json:4]: This file should all V1, V2 & V3 address?

File: rarible/src/exchange-v-1.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import { Transaction } from "../generated/schema";
+// import { isNFT } from "./rarible-helper";
+
+export function getOrCreateTransaction(
+  hash: Bytes,
+  nftAddress: Address,
+  nftId: BigInt
+): Transaction {
+  let entity = Transaction.load(
+    hash.toHexString() +
```
Comment (0xsarvesh) [inline rarible/src/exchange-v-1.ts:13]: Please extract logic of creating id in a different function/variable & reuse same on line 20/

File: rarible/src/exchange-v-1.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        "-" +
+        nftAddress.toHexString() +
+        "-" +
+        nftId.toHexString()
+    );
+  }
+  entity.exchange = "v1";
+  return entity;
+}
+const assetTypeMap = new TypedMap<number, string>();
```
Comment (0xsarvesh) [inline rarible/src/exchange-v-1.ts:31]: Please move this to constant file.

File: rarible/src/exchange-v-1.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    tx.nftId = sellAsset.tokenId;
+    tx.nftAmount = call.inputs.amount;
+    tx.paymentToken = buyAsset.token.toHexString();
+    tx.paymentAmount = paymentAmount;
+    tx.paymentTokenId = buyAsset.tokenId;
+    tx.from = call.inputs.order.key.owner.toHexString();
+    tx.to = call.from.toHexString();
+    tx.blockHeight = call.block.number;
+    tx.hash = call.transaction.hash.toHexString();
+    tx.nft = "sellAsset";
```
Comment (0xsarvesh) [inline rarible/src/exchange-v-1.ts:63]: what is meaning of this hard coded string?

File: rarible/src/exchange-v-1.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+      .div(BigInt.fromI32(10000));
+    let paymentAmount = call.inputs.amount.plus(fee);
+    let tx = getOrCreateTransaction(
+      call.transaction.hash,
+      buyAsset.token,
+      buyAsset.tokenId
+    );
+    tx.nftAddress = buyAsset.token.toHexString();
+    tx.nftId = buyAsset.tokenId;
+    tx.nftAmount = call.inputs.order.buying;
```
Comment (0xsarvesh) [inline rarible/src/exchange-v-1.ts:92]: This transaction object creation logic is duplicated.  Please refactor code & remove duplicity.

File: rarible/src/exchange-v-2.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  if (leftAssetType == ETH || leftAssetType == ERC20) {
+    let tx = getOrCreateTransaction(
+      call.transaction.hash,
+      rightAsset.address,
+      rightAsset.id
+    );
+    tx.hash = call.transaction.hash.toHexString();
+    tx.from = orderRight.maker.toHexString();
+    tx.to = orderLeft.maker.toHexString();
+    tx.nft = "RIGHT";
```
Comment (0xsarvesh) [inline rarible/src/exchange-v-2.ts:49]: what is meaning of this hardcoded string?

File: rarible/schema.graphql [hunk 9]
```diff
@@ -0,0 +1,218 @@
+type Transaction @entity {
+  " hash-nftAddress-id "
+  id: ID!
+  exchange: String
+  hash: String
+  from: String
```
Comment (0xsarvesh) [inline rarible/schema.graphql:6]: This should be address type.

File: rarible/src/exchange-v-2.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    tx.nftAddress = rightAsset.address.toHexString();
+    tx.nftId = rightAsset.id;
+    tx.paymentToken = leftAsset.address.toHexString();
+    let paymentAmount = calculatedTotal(
```
Comment (0xsarvesh) [inline rarible/src/exchange-v-2.ts:53]: Please use const where ever possible.

File: rarible/src/exchange-v-2.ts [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    let tx = getOrCreateTransaction(
+      call.transaction.hash,
+      leftAsset.address,
+      leftAsset.id
+    );
+    tx.hash = call.transaction.hash.toHexString();
+    tx.from = orderLeft.maker.toHexString();
+    tx.to = orderRight.maker.toHexString();
+    tx.nft = "LEFT";
+    tx.nftAddress = leftAsset.address.toHexString();
```
Comment (0xsarvesh) [inline rarible/src/exchange-v-2.ts:83]: Similar comment, remove duplicity of code.

File: rarible/src/modules/abi.ts [hunk 12]
```diff
@@ -0,0 +1,358 @@
+import { Address, BigInt, ByteArray, Bytes, ethereum } from "@graphprotocol/graph-ts";
+import { log } from "matchstick-as";
+
+export namespace abi {
```
Comment (0xsarvesh) [inline rarible/src/modules/abi.ts:4]: It seems, this file has code specific to seaport subgraph. Could you please delete functions which are not used.

File: rarible/src/modules/orders.ts [hunk 13]
```diff
@@ -0,0 +1,220 @@
+import { Address, BigInt, Bytes, log, TypedMap } from "@graphprotocol/graph-ts";
+import { address, integer } from "@protofire/subgraph-toolkit";
+import { shared } from "./shared";
+
+export namespace orders {
```
Comment (0xsarvesh) [inline rarible/src/modules/orders.ts:5]: This file is also specific to seaport.

File: rarible/subgraph.yaml [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+  file: ./schema.graphql
+dataSources:
+  - kind: ethereum
+    name: ExchangeV1
+    network: mainnet
+    source:
+      address: "0xcd4ec7b66fbc029c116ba9ffb3e59351c20b5b06"
+      abi: ExchangeV1
+      startBlock: 11274514 #v1
+      # startBlock: 12506593 #nft <-> token case
```
Comment (0xsarvesh) [inline rarible/subgraph.yaml:12]: Please remove comments.

File: rarible/tests/exchange-v-1-utils.ts [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+  sellValue: BigInt,
+  owner: Address,
+  buyToken: Address,
+  buyTokenId: BigInt,
+  buyValue: BigInt,
+  buyer: Address,
+  amount: BigInt,
+  salt: BigInt
+): Buy {
+  let buyEvent = changetype<Buy>(newMockEvent())
```
Comment (0xsarvesh) [inline rarible/tests/exchange-v-1-utils.ts:21]: Why are we using mocking here?

Comment (0xsarvesh) [reply rarible/tests/exchange-v-1-utils.ts:21]: It should be done in test files.

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]

File: rarible/networks.json (hunk 3 above)
Comment (sharathkrml) [reply rarible/networks.json:4]: Just need to add V2 address right? V3 is implemented as upgrade or V2 only

File: rarible/tests/exchange-v-1-utils.ts (hunk 15 above)
Comment (sharathkrml) [reply rarible/tests/exchange-v-1-utils.ts:21]: it is autogenerated as I did not even touch the tests folder

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**